Implement GitHub Actions for lint and buildAndTest

GleasonK opened the pull request:
Initial implementation of GitHub Actions CI. This changelist implements two actions:

Build & Test:
  - Check out the llvm revision specified in `llvm_version.txt`
  - Build LLVM with ccache enabled (first time cost ~1.5hr, following runs ~10-15min)
  - Build StableHLO with ccache
  - Run `ninja check-stablehlo`
  - Fails if any build or test failures.

Lint:
  - Run `git-clang-format` on the files being submitted.
  - Fails if any files' formatting are updated.

Examples of what the CI page will look like can be seen at: https://github.com/GleasonK/stablehlo/actions

Future work:
- We can consider only running buildAndTest on linux in the future, mac is a bit of a redundancy, but will help us distinguish a real failure from a build configuration issue in GH Actions.
- The lint action worked in the limited testing that I ran on it, but could use a bit more testing. I think that the production repo will be a good way to test if everything is in order.
- Consider any other tests/linting we want to run. If we add python bindings, there is some python formatting and linting that can be added later.

burmako commented on build_tools/github_actions/ci_build_llvm.sh:
```diff
@@ -0,0 +1,61 @@
+# limitations under the License.
+
+# This file is similar to build_mlir.sh, but passes different flags for
+# cacheing in GitHub Actions to improve build speeds.
```
Typo: caching? (Here and below).

burmako commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,25 @@
+      - name: Running clang-format on changed source files
+        run: |
+          /tmp/git-clang-format "${GITHUB_BASE_REF?}" --style=google
+          git diff --exit-code
```
Would it be possible to also run `clang-tidy`, bootstrapping the config from MLIR-HLO to begin with? https://github.com/tensorflow/mlir-hlo/blob/master/.clang-tidy. Doesn't have to block this PR.

GleasonK replied:
Created #61

burmako commented on .github/actions/setup-build/action.yml:
```diff
@@ -0,0 +1,45 @@
+  - name: Install Ninja
+    uses: llvm/actions/install-ninja@55d844821959226fab4911f96f37071c1d4c3268
+
+  # Get LLD
```
Perhaps expand this comment to explain why this is needed?

GleasonK replied:
Added comment "Improves build speed on Linux" - that's the main reason to use LLD ([ref]( https://mlir.llvm.org/getting_started/)). The MHLO build script requires LLD in it's current state so this helps keep those in line, but that's a minor reason since this uses the `ci_build_*` scripts.

burmako commented on .github/actions/setup-build/action.yml:
```diff
@@ -0,0 +1,45 @@
+
+# The setup-build action gets everything needed by buildAndTest into the workspace.
+# This checks out the proper revision of the LLVM repository, installs ninja, and sets
+# up the ccache for faster re-compilation.
```
Consider reformulating this comment to make sure it always stays in sync with the implementaton. For example, it looks like even now it's out of sync because it doesn't call out installation of lld. Maybe drop the second sentence altogether?

burmako commented on .github/actions/setup-build/action.yml:
```diff
@@ -0,0 +1,45 @@
+      LLVM Commit Ref for checkout and build. Used for ccache value and checkout.
+    required: true
+
+
```
How many newlines are idiomatic to separate sections in YAML files? I skimmed some .yml files, and I don't think it's more than 1.

GleasonK replied:
Deleted additional new line. I tried some searching for YMAL conventions and it looks like a mix of 1 line and 0 lines.

burmako commented on .github/actions/setup-build/action.yml:
```diff
@@ -0,0 +1,45 @@
+name: "Setup build environment (ninja, ccache, llvm, lld)"
+
+inputs:
+  llvm-ref:
```
At the moment, we're using multiple ways of referring to the same thing: 1) llvm-ref, 2) "LLVM Commit Ref", 3) "checkout llvm at revision", 4) llvm_version.txt. How about we standardize on one name, e.g. 1) llvm-version, 2) "LLVM version", 3) "checkout llvm at version"? Same for the files below.

burmako commented on .github/workflows/buildAndTest.yml:
```diff
@@ -0,0 +1,71 @@
+  push:
+    branches: [ main ]
+  schedule:
+    # Always regenerate once every day
```
Wdym by "regenerate"?

GleasonK replied:
Modified comment. "Run this workflow once a day." 

Regenerate doesn't make sense unless we're publishing releases or something.

burmako commented on .github/workflows/buildAndTest.yml:
```diff
@@ -0,0 +1,71 @@
+
+# Build using CMake and run tests.
+# Use Cached LLVM to improve build times.
+# Current setup only builds on linux.
```
Is this comment still up-to-date given that we say `os-arch: maci-x86_64` below?

GleasonK replied:
deleted.

burmako commented on .github/workflows/buildAndTest.yml:
```diff
@@ -0,0 +1,71 @@
+
+# Build using CMake and run tests.
+# Use Cached LLVM to improve build times.
+# Current setup only builds on linux.
```
Also, what about Windows? How hard would it be to implement it as well? This definitely doesn't have to block this CL - can just open a ticket for that if it's realistic, or we can drop this topic if it's not.

GleasonK replied:
Would require some new scripts since most of our workflows are bash. None of the other MLIR OSS projects seem to have actions for windows. Still maybe worth considering?

burmako replied:
"None of the other MLIR OSS projects seem to have actions for windows". Good to know! Let's ignore for now then.

burmako commented on .github/workflows/buildAndTest.yml:
```diff
@@ -0,0 +1,71 @@
+    steps:
+
+    # Check out stablehlo repo
+    - name: Checkout stablehlo
```
"StableHLO", since "LLVM" is capitalized in the similar context?

burmako commented on .github/workflows/buildAndTest.yml:
```diff
@@ -0,0 +1,71 @@
+          - os-arch: maci-x86_64 
+            os: macos-latest
+            llvm-build-dir: "$GITHUB_WORKSPACE/llvm-build-maci"
+            stablehlo-build-dir: "$GITHUB_WORKSPACE/build-maci"
```
Why are Linux and macOS directories different?

GleasonK replied:
Removed. I think I was dodging the risk that there may be some sort of NFS thing going on with shared directories between the two platforms, but it doesn't seem like that's the case with GH actions.

burmako commented on build_tools/github_actions/ci_build_llvm.sh:
```diff
@@ -0,0 +1,61 @@
+  FLAGS="$LLD_FLAG"
+else
+  echo "lld not found, using default linker"
+fi
```
In what cases do we not have `lld` in our CI environment?

GleasonK replied:
Currently mac until I can figure out a good way to install lld on mac in the ci environment.

burmako replied:
What is the value that macOS support brings to the table? From what I can see, it complicates the configuration and the scripts, and it'll probably slow down CI overall, but I don't think I have a good understanding of what we're getting in return.

GleasonK replied:
Linux was giving trouble early on, mac was not. Initially I was using it to sanity check - i.e. if Linux failed but Mac passed, it probably had more to do with the GH Linux environment than the code change.

That said, I think that was largely due to the combination of caching and different experiments I was doing with config. I cleared all caches a little over a week ago and haven't had any Linux issues since then, so we're probably good to get rid of Mac.

burmako commented on build_tools/github_actions/ci_build_llvm.sh:
```diff
@@ -0,0 +1,61 @@
+  -DLLVM_INCLUDE_TESTS=OFF \
+  -DLLVM_INCLUDE_TOOLS=ON \
+  -DLLVM_INSTALL_UTILS=ON \
+  -DLLVM_TARGETS_TO_BUILD="X86;NVPTX;AMDGPU"
```
Consider ordering CMake flags similarly to https://github.com/openxla/stablehlo/blob/9ca259d5092e9cf2c1fa0788a470df6a4fc95f0a/build_tools/build_mlir.sh#L40-L52 to make these files easier to compare side-by-side, e.g. 

```
cmake -GNinja \
  "-H$LLVM_SRC_DIR/llvm" \
  "-B$LLVM_BUILD_DIR" \
  -DLLVM_INSTALL_UTILS=ON \
  "$LLVM_ENABLE_LLD" \
  -DLLVM_ENABLE_PROJECTS=mlir \
  -DLLVM_TARGETS_TO_BUILD="X86;NVPTX;AMDGPU" \
  -DLLVM_INCLUDE_TOOLS=ON \
  -DLLVM_ENABLE_BINDINGS=OFF \
  -DLLVM_BUILD_TOOLS=OFF \
  -DLLVM_INCLUDE_TESTS=OFF \
  -DCMAKE_BUILD_TYPE=RelWithDebInfo \
  -DLLVM_ENABLE_ASSERTIONS=On \
  -DCMAKE_CXX_COMPILER=clang++ \
  -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
  -DCMAKE_C_COMPILER=clang \
  -DCMAKE_C_COMPILER_LAUNCHER=ccache
```

GleasonK replied:
The flags in this file are in alphabetical order (sorted by vim). Made the suggested change to make these more consistent, since it seems like alphabetical isn't a convention for cmake flags.

burmako commented on build_tools/github_actions/ci_build_stablehlo.sh:
```diff
@@ -0,0 +1,55 @@
+  -DMLIR_DIR="$LLVM_BUILD_DIR/lib/cmake/mlir"
+
+cd "$STABLEHLO_BUILD_DIR"
+ninja check-stablehlo
```
How hard would it be to also build and test Python bindings as described in https://github.com/openxla/stablehlo/tree/main/build_tools#python-api? This doesn't have to block this PR, but then we'll need a ticket, because we'll definitely want to have Python bindings in CI.

GleasonK replied:
Created #62

GleasonK commented:
Addressed feedback. Actions job with the new changes is running on https://github.com/GleasonK/stablehlo/actions. Let's wait for that before any accept/merge is made.

burmako commented on .github/workflows/buildAndTest.yml:
```diff
@@ -0,0 +1,65 @@
+      run: |
+        ./build_tools/github_actions/ci_build_llvm.sh "$GITHUB_WORKSPACE/llvm-project/" "$GITHUB_WORKSPACE/llvm-build"
+
+    # Configure for linux
```
Looks like an out of date / copypasted comment?

GleasonK commented:
Edit: GH Actions job passed. Only took 4m 53s with caching, not bad.

burmako commented on .github/actions/setup-build/action.yml:
```diff
@@ -0,0 +1,42 @@
+
+  # Get LLD - Improves build speed on Linux
+  - name: Install LLD
+    if: "runner.os == 'Linux'"
```
`if: "runner.os == 'Linux'"` => this can be deleted as well now?

burmako reviewed (approved):
Please also update the PR description to reword "We can consider only running buildAndTest on linux in the future, mac is a bit of a redundancy, but will help us distinguish a real failure from a build configuration issue in GH Actions".
